src/mpid/ch4/shm/posix/eager/fbox/fbox_init.c

```diff
@@ -125,18 +136,215 @@ int MPIDI_POSIX_eager_init(int rank, int size)
+        mpi_errno =
+            MPIDU_shm_seg_alloc(MPIDI_POSIX_global.num_local * num_procs[i] *
+                                sizeof(MPIDI_POSIX_fastbox_t), (void **) &fastboxes_p[i],
+                                MPL_MEM_SHM);
+        if (mpi_errno)
+            MPIR_ERR_POP(mpi_errno);
+
+        /* Actually allocate the segment and assign regions to the pointers */
+        mpi_errno = MPIDU_shm_seg_commit(&local_shm_area[i].memory,
+                                         &local_shm_area[i].barrier,
+                                         MPIDI_POSIX_global.num_local,
+                                         MPIDI_POSIX_global.my_local_rank,
+                                         MPIDI_POSIX_global.local_rank_0, rank, MPL_MEM_SHM);
+        if (mpi_errno)
+            MPIR_ERR_POP(mpi_errno);
+
+        /* Increment number of shared memory areas allocated */
+        local_shm_area_num++;
+    }
+
+    MPIDI_POSIX_global.barrier = local_shm_area[0].barrier;
 
+    /* Allocate table of pointers to fastboxes */
+    MPIR_CHKPMEM_MALLOC(MPIDI_POSIX_eager_fbox_control_global.mailboxes.in,
+                        MPIDI_POSIX_fastbox_t **,
+                        MPIDI_POSIX_global.num_local * sizeof(MPIDI_POSIX_fastbox_t *), mpi_errno,
+                        "fastboxes", MPL_MEM_SHM);
+    MPIR_CHKPMEM_MALLOC(MPIDI_POSIX_eager_fbox_control_global.mailboxes.out,
+                        MPIDI_POSIX_fastbox_t **,
+                        MPIDI_POSIX_global.num_local * sizeof(MPIDI_POSIX_fastbox_t *), mpi_errno,
+                        "fastboxes", MPL_MEM_SHM);
+
+#define SEGMENT_INDEX_IMPL(p_domain_id) ({        \
+    int i_;                                       \
+    do {                                          \
+        for (i_ = 0; i_ < n_domains; i_++) {      \
+            if (s_domain_ids[i_] == p_domain_id)  \
+                break;                            \
+        }                                         \
+    } while (0);                                  \
+    i_;                                           \
+})
+#define SEGMENT_INDEX(receiver) (SEGMENT_INDEX_IMPL(topo->p_domain_ids[receiver]))
+#define PROCESS_INDEX(sender, receiver) (((topo->s_ranks[receiver]) * (MPIDI_POSIX_global.num_local)) + (sender))
+
+    /* Fill in fastbox tables */
+    int segment_idx, process_idx;
+    for (i = 0; i < MPIDI_POSIX_global.num_local; i++) {
+        segment_idx = SEGMENT_INDEX(MPIDI_POSIX_global.my_local_rank);
+        process_idx = PROCESS_INDEX(i, MPIDI_POSIX_global.my_local_rank);
+        MPIDI_POSIX_eager_fbox_control_global.mailboxes.in[i] =
+            &fastboxes_p[segment_idx][process_idx];
+
+        segment_idx = SEGMENT_INDEX(i);
+        process_idx = PROCESS_INDEX(MPIDI_POSIX_global.my_local_rank, i);
+        MPIDI_POSIX_eager_fbox_control_global.mailboxes.out[i] =
+            &fastboxes_p[segment_idx][process_idx];
+    }
+
```

Comment:
@pavanbalaji actually, I have to make a correction. The topology information is needed at the shmmod level, outside the memory allocation function, because the shmmod has to do the mapping of `mailboxes` to `fastboxes` inside shared memory segments. Same problem will be there for shared memory windows. I can get rid of the freeing of the initial segment if I do the topology aware allocation at POSIX shmmod init time but in that case I will need to use a shared memory segment for the topology information exchange among processes and this will add code duplication. Doing everything in the memory allocation function as you suggested (i.e., `MPIDU_shm_seg_commit`) will still require this to leak topology information to allow other components to do the appropriate mappings.

Reply:
Depends on what the abstraction is.  If the abstraction is: "here are 4 processes; allocate X bytes per process", then the fact these bytes are used for fastbox or something else is irrelevant.

Reply:
If the allocation function is going to be responsible for allocating and binding memory then the abstraction should be: "here are 4 processes, their memory binding policies (defined by the user, e.g., through CVARS), and their mapping to the memory of the node (defined by the user through `mpirun`, i.e., `--bind-to`), give me back the appropriate number of segments bound to the appropriate memory nodes". Then you get back N-segments bound as requested. That should be it, mapping processes to memory segments is not the job of the allocation function.

Reply:
Can you present your abstraction first?  It's hard to argue without knowing that.

Reply:
This is a rough idea, I may still need to break this down into alloc + commit as CH3 still needs it to allocate multiple objects in the same segment.

```
int MPIDI_SHM_alloc(int num_local,
                    int local_rank,
                    int local_proc_0,
                    int rank,                       /* input:  first four fields as for MPIDU_shm_seg_commit */
                    int mem_id,                     /* input:  memory id to which bind memory, obtained through
                                                     *         MPIR_Node_get_covering_obj_by_type(memtype) +
                                                     *         MPIR_Node_get_obj_index */
                    MPIDI_shm_bind_info_t * info,   /* input:  bind policy, e.g., bind, interleave, and flags, e.g., map by DDR or MCDRAM set */
                    MPIDU_shm_seg_t ** memory,      /* return: shared memory segment descriptors [0..num_seg-1] */
                    MPIDU_shm_barrier_t ** barrier, /* return: shared memory barrier (only one) */
                    int * num_seg,                  /* return: number of allocated segments */
                    void **ptr,                     /* return: pointer to process region in segment */
                    int **mapping                   /* return: array mapping processes to segment ids [0..num_local-1]*/
                    );
```

Reply:
Why do you need `MPIDI_shm_bind_info_t`?

Reply:
The memory id is just an integer. Right now all memory types are numa nodes but this might change and you may have MCDRAM and DDR with same memory id. The binding info tells you how to bind the memory, e.g., BYDDRSET or BYMCDRAM set. For now these are simply converted to `HWLOC_MEMBIND_BYNODESET`. Additionally, if the policy is `INTERLEAVE` `mem_id` will be discarded as segments will be spread out across all memory nodes.

Reply:
My question is, why does the caller need to specify that?

Reply:
How would the alloc function know what is the memory binding policy for the requested allocation? Only the caller knows

Reply:
How does the caller know?  Isn't it simply getting that info from a CVAR?

Reply:
The caller has to get that from `MPIDI_shm_bind_get(obj_type, info)` (which will parse the CVAR for the `obj_type`) as each device can have its own objects. Then pass it to memory allocation to get memory back. Is this wrong?

Reply:
`each device can have its own objects`

What does this mean?  I tend to think of fastbox and cells as objects.  Here we are talking about MCDRAM and DRAM -- why are these specific to a device?  They are a characteristic of the hardware platform.

Reply:
MCDRAM and DRAM are not specific to device. Sorry I may have been unclear. What I meant is that every device can define what memory objects it accepts memory binding policies for, e.g., `fastboxes` or `iqueues`, by declaring CVARs for them. The CVARs will contain all the memory binding information needed to allocate the objects and bind it, e.g., `MEMTYPE` (DDR|MCDRAM|OTHERS), `POLICY` (BIND|INTERLEAVE) and binding `FLAGS` (STRICT).

Reply:
`every device can define what memory objects it accepts memory binding policies for`

Why?  How is that better than the CVARs that allow the user to control it?

Reply:
Sure we can make the user decide the objects too not just the policy for them. But then we are back to the original issue. Either `MPIDI_shm_bind_info_t` or the CVAR itself is needed to give the alloc function the context of what object it is working on.

Reply:
Why is the CVAR defined by the device instead of the buffer allocation class?

Reply:
If you prefer that sure. Basically if I understand correctly you want the alloc function to get the binding information from the CVAR itself and the user should just tell the alloc function what object is being allocated?